source/h2_frames.c

```diff
@@ -0,0 +1,1618 @@
+        goto read_error;
+    }
+
+    if (decoder->flags) {
+        goto protocol_error;
+    }
+
+    /* Initialize the frame */
+    frame->header = decoder->header;
+
+    /* Decode the priority settings */
+    if (s_frame_priority_settings_decode(&frame->priority, decoder)) {
+        goto read_error;
+    }
+    AWS_FATAL_ASSERT(decoder->payload.len == 0);
+
+    return AWS_OP_SUCCESS;
+
+read_error:
+    *decoder = decoder_init;
+    return aws_raise_error(AWS_H2_ERR_FRAME_SIZE_ERROR);
+
+protocol_error:
+    *decoder = decoder_init;
+    return aws_raise_error(AWS_H2_ERR_PROTOCOL_ERROR);
+}
+
+/***********************************************************************************************************************
+ * RST_STREAM
+ **********************************************************************************************************************/
+static const size_t s_frame_rst_stream_length = 4;
+
+int aws_h2_frame_rst_stream_init(struct aws_h2_frame_rst_stream *frame, struct aws_allocator *allocator) {
+    (void)allocator;
+
+    AWS_ZERO_STRUCT(*frame);
+    frame->header.type = AWS_H2_FRAME_T_RST_STREAM;
+
+    return AWS_OP_SUCCESS;
+}
+void aws_h2_frame_rst_stream_clean_up(struct aws_h2_frame_rst_stream *frame) {
+    AWS_PRECONDITION(frame);
+
+    (void)frame;
+}
+
+int aws_h2_frame_rst_stream_encode(
+    struct aws_h2_frame_rst_stream *frame,
+    struct aws_h2_frame_encoder *encoder,
+    struct aws_byte_buf *output) {
+    AWS_PRECONDITION(frame);
+    AWS_PRECONDITION(encoder);
+    AWS_PRECONDITION(output);
+
+    (void)encoder;
+
+    const size_t output_init_len = output->len;
+
+    /* Write the header data */
+    if (s_frame_header_encode(&frame->header, s_frame_rst_stream_length, 0, output)) {
+        goto write_error;
+    }
+
+    /* Write the error_code */
+    if (!aws_byte_buf_write_be32(output, frame->error_code)) {
+        goto write_error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+write_error:
+    output->len = output_init_len;
+    return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+}
+int aws_h2_frame_rst_stream_decode(struct aws_h2_frame_rst_stream *frame, struct aws_h2_frame_decoder *decoder) {
+    AWS_PRECONDITION(frame);
+    AWS_PRECONDITION(decoder);
+
+    if (aws_h2_frame_rst_stream_init(frame, decoder->allocator)) {
+        return AWS_OP_ERR;
+    }
+
+    AWS_ASSERT(decoder->header.type == AWS_H2_FRAME_T_RST_STREAM);
+
+    const struct aws_h2_frame_decoder decoder_init = *decoder;
+
+    if (decoder->payload.len != s_frame_rst_stream_length) {
+        goto read_error;
+    }
+
+    if (decoder->flags) {
+        goto protocol_error;
+    }
+
+    /* Initialize the frame */
+    frame->header = decoder->header;
+
+    /* Decode the priority settings */
+    uint32_t error_code = 0;
+    if (!aws_byte_cursor_read_be32(&decoder->payload, &error_code)) {
+        goto read_error;
+    }
+    frame->error_code = (enum aws_h2_error_codes)error_code;
+    AWS_FATAL_ASSERT(decoder->payload.len == 0);
```

Comment:
Shouldn't this just be a protocol error?

Reply:
Nah, sanity check. If the size is wrong line 919 will return a FRAME_SIZE_ERROR.